### PR TITLE
os/bluestore: Make spending cache budget for data cache more aggresive

### PR DESCRIPTION
**Pull request by aclamk**
This change takes into account that different shards can vastly vary in usage.
Reason is that different shards can cache pgs of different pools.
Now shards that need more space are assigned more at expense of less used shards.

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -4155,8 +4155,19 @@ void BlueStore::MempoolThread::_resize_shards(bool interval_stats)
   for (auto i : store->onode_cache_shards) {
     i->set_max(max_shard_onodes);
   }
+  //sum is in data_used
+  uint64_t actual_trim = 0;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:4159]: unused

Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:4159]: and max_shard_buffer declaration above to be removed

**Comment by ifed01** [comment]
Generally the ideal looks good but may be we should calculate trim size per shard  based on this shard's contribution into overall volume. E,g, 
shard1 has 900 bytes and shard2 has 100 bytes. Hence 90% contribution for s1 and 10% for s2
And one needs to trim 500 bytes totals. 
Current implementation will result in shard1 = 650 (=900 - 500/2) and shard2 = 0.
Instead we can estimate trim size from shard1 at 450 (0.9 * 500) and shard2 = 50(0.1 *500).
Hence leaving another 450 & 50 bytes in each shard respectively.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: @ifed01 thanks!

**Comment by aclamk** [comment]
> Generally the idea looks good but may be we should calculate trim size per shard based on this shard's contribution into overall volume. E,g,
> shard1 has 900 bytes and shard2 has 100 bytes. Hence 90% contribution for s1 and 10% for s2
> And one needs to trim 500 bytes totals.
> Current implementation will result in shard1 = 650 (=900 - 500/2) and shard2 = 0.
> Instead we can estimate trim size from shard1 at 450 (0.9 * 500) and shard2 = 50(0.1 *500).
> Hence leaving another 450 & 50 bytes in each shard respectively.
@ifed01 This is very fair observation, but I never observed shards use full allocated size. 
Please take a look at https://docs.google.com/spreadsheets/d/1V6iDMBBOW3t0sqEqXoOsmzS0BhA10K4wdBrH7dL1TVk

Still, I just checked it, and your point is more then valid. Some shards are assigned very low values, and it will be almost impossible for those shards to recover from this state.
`2020-10-13T10:38:40.220-0400 7f2d4f2b0700 30 bluestore.MempoolThread(0x55b7025529c0) _resize_shards cache shards allocation: 0/7808  0/7808  56668160/56675968  52133888/52141696  0/7808  0/7808  44023808/44031616  0/7808  51752960/51760768  0/7808  0/7808  60043264/60051072  0/7808  0/7808  0/7808  56823808/56839808  62939136/62946944  77619200/77627008  0/7808  55685120/55692928  63467520/63471232  0/7808  53149696/53161600  58368000/58375808  91348992/91356800  50081792/50089600  0/7808  56258560/56266368  95145984/95153792  54198272/54201984  55365632/55377536  53907456/53915264`

**Comment by markhpc** [comment]
Another route that might be interesting to try pursuing is to have each cache shard compete independently in the prioritycache.  The idea being that with the introduction of age binning we can directly assign memory to each cache shard and avoid the extra complexity of aggregating the requests and breaking down the distribution. Given that we already have much of that code in place I'm not sure if it's actually a better option though.

**Pull request closed without merging**